src/topics/create.ts

```diff
@@ -0,0 +1,380 @@
+import _ from 'lodash';
+
+import db from '../database';
+import utils from '../utils';
+import slugify from '../slugify';
+import plugins from '../plugins';
+import analytics from '../analytics';
+import user from '../user';
+import meta from '../meta';
+import posts from '../posts';
+import privileges from '../privileges';
+import categories from '../categories';
+import translator from '../translator';
+
+import { TopicData, TopicFields, PostObjectPartial, SettingsObject, TopicObject, PostObject, TopicAndPostData, UserObjectSlim } from '../types';
+
+async function guestHandleValid(data) {
+    // The next line calls a function in a module that has not been updated to TS yet
+    // eslint-disable-next-line @typescript-eslint/no-unsafe-member-access, @typescript-eslint/no-unsafe-call
+    if (meta.config.allowGuestHandles && parseInt(data.uid as string, 10) === 0 && data.handle) {
+        // The next line calls a function in a module that has not been updated to TS yet
+        // eslint-disable-next-line @typescript-eslint/no-unsafe-member-access, @typescript-eslint/no-unsafe-call
+        if (data.handle.length > meta.config.maximumUsernameLength) {
+            throw new Error('[[error:guest-handle-invalid]]');
+        }
+        // The next line calls a function in a module that has not been updated to TS yet
+        // eslint-disable-next-line @typescript-eslint/no-unsafe-member-access, @typescript-eslint/no-unsafe-call
+        const exists = await user.existsBySlug(slugify(data.handle));
+        if (exists) {
+            throw new Error('[[error:username-taken]]');
+        }
+    }
+}
+
+function check(item: string, min: string, max: string, minError: string, maxError: string) {
+    // Trim and remove HTML (latter for composers that send in HTML, like redactor)
+    if (typeof item === 'string') {
+        item = utils.stripHTMLTags(item).trim();
+    }
+
+    if (item === null || item === undefined || item.length < parseInt(min, 10)) {
+        throw new Error(`[[error:${minError}, ${min}]]`);
+    } else if (item.length > parseInt(max, 10)) {
+        throw new Error(`[[error:${maxError}, ${max}]]`);
+    }
+}
+
+
+async function canReply(data: TopicData, topicData: TopicObject) {
+    if (!topicData) {
+        throw new Error('[[error:no-topic]]');
+    }
+    const { tid, uid } = data;
+    const { cid, deleted, locked, scheduled } = topicData;
+
+    const [canReply, canSchedule, isAdminOrMod]: boolean[] = await Promise.all([
+        // The next line calls a function in a module that has not been updated to TS yet
+        // eslint-disable-next-line @typescript-eslint/no-unsafe-member-access, @typescript-eslint/no-unsafe-call
```

Comment:
nit: could you add a note somewhere in the PR description tagging jake to remove these lines once privileges has been translated